### PR TITLE
Fix pre-volta Desul atomics bug

### PR DESCRIPTION
**Pull request by dalg24**
Tentative fix for #4360

### CODE REVIEW

**Comment by ndellingwood** [comment]
Tested this on the failing machine for #4360 , build completed :)

```
git fetch origin pull/4367/head:pr-4367
git checkout pr-4367

module load sems-env sems-cmake/3.17.1 sems-clang/7.0.1 sems-cuda/9.2
$KOKKOS_PATH/generate_makefile.bash --with-devices=Cuda,OpenMP --arch=Kepler35 --compiler=clang++ --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized " --cxxstandard="14" --with-cuda=/projects/sems/install/rhel7-x86_64/sems/compiler/cuda/9.2/base --kokkos-path=$KOKKOS_PATH --with-options=disable_deprecated_code --no-examples
```

**Review by ndellingwood — APPROVED** [review]
lgtm

**Comment by ndellingwood** [comment]
Passed unit testing in configuration mentioned above https://github.com/kokkos/kokkos/pull/4367#issuecomment-929543006

```
[ndellin@kokkos-dev NightlyFail]$ ctest
Test project /ascldap/users/ndellin/kokkos/testing/NightlyFail
      Start  1: KokkosCore_UnitTest_OpenMP
 1/52 Test  #1: KokkosCore_UnitTest_OpenMP ...................   Passed   33.22 sec
      Start  2: KokkosCore_UnitTest_OpenMPInterOp
 2/52 Test  #2: KokkosCore_UnitTest_OpenMPInterOp ............   Passed    5.97 sec
      Start  3: KokkosCore_UnitTest_OpenMPGraph
 3/52 Test  #3: KokkosCore_UnitTest_OpenMPGraph ..............   Passed    6.83 sec
      Start  4: KokkosCore_UnitTest_Cuda1
 4/52 Test  #4: KokkosCore_UnitTest_Cuda1 ....................   Passed  114.64 sec
      Start  5: KokkosCore_UnitTest_Cuda2
 5/52 Test  #5: KokkosCore_UnitTest_Cuda2 ....................   Passed   21.40 sec
      Start  6: KokkosCore_UnitTest_Cuda3
 6/52 Test  #6: KokkosCore_UnitTest_Cuda3 ....................   Passed  148.51 sec
      Start  7: KokkosCore_UnitTest_CudaTimingBased
 7/52 Test  #7: KokkosCore_UnitTest_CudaTimingBased ..........   Passed    7.25 sec
      Start  8: KokkosCore_UnitTest_CudaInterOpInit
 8/52 Test  #8: KokkosCore_UnitTest_CudaInterOpInit ..........   Passed    6.70 sec
      Start  9: KokkosCore_UnitTest_CudaInterOpStreams
 9/52 Test  #9: KokkosCore_UnitTest_CudaInterOpStreams .......   Passed    6.50 sec
      Start 10: KokkosCore_UnitTest_CudaGraph
10/52 Test #10: KokkosCore_UnitTest_CudaGraph ................   Passed    6.73 sec
      Start 11: KokkosCore_UnitTest_Default
11/52 Test #11: KokkosCore_UnitTest_Default ..................   Passed    6.82 sec
      Start 12: KokkosCore_UnitTest_PushFinalizeHook
12/52 Test #12: KokkosCore_UnitTest_PushFinalizeHook .........   Passed    6.46 sec
      Start 13: KokkosCore_UnitTest_Develop
13/52 Test #13: KokkosCore_UnitTest_Develop ..................   Passed    6.67 sec
      Start 14: KokkosCore_UnitTest_LogicalSpaces
14/52 Test #14: KokkosCore_UnitTest_LogicalSpaces ............   Passed    7.92 sec
      Start 15: KokkosCore_UnitTest_KokkosP
15/52 Test #15: KokkosCore_UnitTest_KokkosP ..................   Passed    6.72 sec
      Start 16: KokkosCore_ProfilingTestLibraryLoadHelp
16/52 Test #16: KokkosCore_ProfilingTestLibraryLoadHelp ......   Passed    6.61 sec
      Start 17: KokkosCore_ProfilingTestLibraryCmdLineHelp
17/52 Test #17: KokkosCore_ProfilingTestLibraryCmdLineHelp ...   Passed    7.00 sec
      Start 18: KokkosCore_ProfilingTestLibraryLoad
18/52 Test #18: KokkosCore_ProfilingTestLibraryLoad ..........   Passed    6.55 sec
      Start 19: KokkosCore_ProfilingTestLibraryCmdLine
19/52 Test #19: KokkosCore_ProfilingTestLibraryCmdLine .......   Passed    6.42 sec
      Start 20: KokkosCore_UnitTest_StackTraceTest
20/52 Test #20: KokkosCore_UnitTest_StackTraceTest ...........   Passed    8.04 sec
      Start 21: KokkosCore_UnitTest_DefaultInit_1
21/52 Test #21: KokkosCore_UnitTest_DefaultInit_1 ............   Passed    6.16 sec
      Start 22: KokkosCore_UnitTest_DefaultInit_2
22/52 Test #22: KokkosCore_UnitTest_DefaultInit_2 ............   Passed    6.47 sec
      Start 23: KokkosCore_UnitTest_DefaultInit_3
23/52 Test #23: KokkosCore_UnitTest_DefaultInit_3 ............   Passed    6.53 sec
      Start 24: KokkosCore_UnitTest_DefaultInit_4
24/52 Test #24: KokkosCore_UnitTest_DefaultInit_4 ............   Passed    6.62 sec
      Start 25: KokkosCore_UnitTest_DefaultInit_5
25/52 Test #25: KokkosCore_UnitTest_DefaultInit_5 ............   Passed    6.56 sec
      Start 26: KokkosCore_UnitTest_DefaultInit_6
26/52 Test #26: KokkosCore_UnitTest_DefaultInit_6 ............   Passed    6.70 sec
      Start 27: KokkosCore_UnitTest_DefaultInit_7
27/52 Test #27: KokkosCore_UnitTest_DefaultInit_7 ............   Passed    6.69 sec
      Start 28: KokkosCore_UnitTest_DefaultInit_8
28/52 Test #28: KokkosCore_UnitTest_DefaultInit_8 ............   Passed    6.66 sec
      Start 29: KokkosCore_UnitTest_DefaultInit_9
29/52 Test #29: KokkosCore_UnitTest_DefaultInit_9 ............   Passed    6.98 sec
      Start 30: KokkosCore_UnitTest_DefaultInit_10
30/52 Test #30: KokkosCore_UnitTest_DefaultInit_10 ...........   Passed    6.89 sec
      Start 31: KokkosCore_UnitTest_DefaultInit_11
31/52 Test #31: KokkosCore_UnitTest_DefaultInit_11 ...........   Passed    6.62 sec
      Start 32: KokkosCore_UnitTest_DefaultInit_12
32/52 Test #32: KokkosCore_UnitTest_DefaultInit_12 ...........   Passed    5.56 sec
      Start 33: KokkosCore_UnitTest_DefaultInit_13
33/52 Test #33: KokkosCore_UnitTest_DefaultInit_13 ...........   Passed    6.45 sec
      Start 34: KokkosCore_UnitTest_DefaultInit_14
34/52 Test #34: KokkosCore_UnitTest_DefaultInit_14 ...........   Passed    7.02 sec
      Start 35: KokkosCore_UnitTest_DefaultInit_15
35/52 Test #35: KokkosCore_UnitTest_DefaultInit_15 ...........   Passed    6.78 sec
      Start 36: KokkosCore_UnitTest_DefaultInit_16
36/52 Test #36: KokkosCore_UnitTest_DefaultInit_16 ...........   Passed    6.51 sec
      Start 37: KokkosCore_UnitTest_DefaultInit_17
37/52 Test #37: KokkosCore_UnitTest_DefaultInit_17 ...........   Passed    7.00 sec
      Start 38: KokkosCore_UnitTest_DefaultInit_18
38/52 Test #38: KokkosCore_UnitTest_DefaultInit_18 ...........   Passed    7.00 sec
      Start 39: KokkosCore_IncrementalTest_OPENMP
39/52 Test #39: KokkosCore_IncrementalTest_OPENMP ............   Passed    7.03 sec
      Start 40: KokkosCore_IncrementalTest_CUDA
40/52 Test #40: KokkosCore_IncrementalTest_CUDA ..............   Passed   44.17 sec
      Start 41: KokkosCore_UnitTest_CTestDevice
41/52 Test #41: KokkosCore_UnitTest_CTestDevice ..............   Passed    0.02 sec
      Start 42: KokkosCore_UnitTest_CMakePassCmdLineArgs0
42/52 Test #42: KokkosCore_UnitTest_CMakePassCmdLineArgs0 ....   Passed    0.00 sec
      Start 43: KokkosCore_PerfTestExec
43/52 Test #43: KokkosCore_PerfTestExec ......................   Passed   32.07 sec
      Start 44: KokkosCore_PerformanceTest_Atomic
44/52 Test #44: KokkosCore_PerformanceTest_Atomic ............   Passed   59.99 sec
      Start 45: KokkosCore_PerformanceTest_Atomic_MinMax
45/52 Test #45: KokkosCore_PerformanceTest_Atomic_MinMax .....   Passed    6.58 sec
      Start 46: KokkosCore_PerformanceTest_Mempool
46/52 Test #46: KokkosCore_PerformanceTest_Mempool ...........   Passed    5.37 sec
      Start 47: KokkosCore_PerformanceTest_TaskDag
47/52 Test #47: KokkosCore_PerformanceTest_TaskDag ...........   Passed    0.01 sec
      Start 48: KokkosContainers_UnitTest_OpenMP
48/52 Test #48: KokkosContainers_UnitTest_OpenMP .............   Passed   89.67 sec
      Start 49: KokkosContainers_UnitTest_Cuda
49/52 Test #49: KokkosContainers_UnitTest_Cuda ...............   Passed   36.52 sec
      Start 50: KokkosContainers_PerformanceTest_OpenMP
50/52 Test #50: KokkosContainers_PerformanceTest_OpenMP ......   Passed   13.51 sec
      Start 51: KokkosContainers_PerformanceTest_Cuda
51/52 Test #51: KokkosContainers_PerformanceTest_Cuda ........   Passed    9.10 sec
      Start 52: KokkosAlgorithms_UnitTest
52/52 Test #52: KokkosAlgorithms_UnitTest ....................   Passed   39.81 sec

100% tests passed, 0 tests failed out of 52

Total Test time (real) = 890.06 sec
```

File: core/src/desul/atomics/CUDA.hpp [hunk 1]
```diff
@@ -453,6 +450,14 @@ namespace desul {
     return Impl::atomic_fetch_oper(Impl::AddOper<long, const long>(), dest, val, order, scope);
   }
   __device__ inline
+  unsigned long atomic_fetch_add(unsigned long* const dest, unsigned long val, MemoryOrderRelaxed order, MemoryScopeDevice scope) {
+    return Impl::atomic_fetch_oper(Impl::AddOper<unsigned long, const unsigned long>(), dest, val, order, scope);
+  }
+  __device__ inline
+  double atomic_fetch_add(double* const dest, double val, MemoryOrderRelaxed order, MemoryScopeDevice scope) {
+    return Impl::atomic_fetch_oper(Impl::AddOper<double, const double>(), dest, val, order, scope);
+  }
```
Comment (dalg24) [inline core/src/desul/atomics/CUDA.hpp:459]: I think this one (and the corresponding `atomic_fetch_add -> double`) might be only needed for Pascal

**Comment by dalg24** [comment]
Retest this please

File: core/src/desul/atomics/CUDA.hpp (hunk 1 above)
Comment (dalg24) [reply core/src/desul/atomics/CUDA.hpp:459]: Broke Clang build on Volta

Comment (dalg24) [reply core/src/desul/atomics/CUDA.hpp:459]: @crtrott @ndellingwood do we have a Clang+CUDA nightly build on Pascal?

**Review by crtrott — APPROVED** [review]

**Pull request merged**